### PR TITLE
feat: add saml request to link to sessions

### LINKED ISSUE #6053: Session API: SAML Create Session
As a developer I am able to create an SAML session, so I can build my own login on top of ZITADEL.

**Acceptance Criteria**
- [ ] SAML Endpoints are implemented as in the current version
- [ ] Endpoint: SAML metadata
- [ ] Endpoint: certificate
- [ ] Endpoint: sso
- [ ] Logout
- [ ] Testing

**Additional Information**
- Separat Auth Request for SAML
- Integration into session of the SAML Auth Request

### PR DESCRIPTION
**Pull request by stebenz**
# Which Problems Are Solved

It is currently not possible to use SAML with the Session API.

# How the Problems Are Solved

Add SAML service, to get and resolve SAML requests.
Add SAML session and SAML request aggregate, which can be linked to the Session to get back a SAMLResponse from the API directly. 

# Additional Changes

Update of dependency zitadel/saml to provide all functionality for handling of SAML requests and responses.

# Additional Context

Closes #6053

### CODE REVIEW

**Review by livio-a — CHANGES_REQUESTED** [review]
can you please document the flow analog to https://zitadel.com/docs/guides/integrate/login-ui/oidc-standard
or create a follow up issue for it

File: proto/zitadel/settings/v2/login_settings.proto [hunk 1]
```diff
@@ -134,6 +134,7 @@ message IdentityProvider {
   string id = 1;
   string name = 2;
   IdentityProviderType type = 3;
+  bool is_linking_allowed = 4;
```
[the remark this answers is not in the document]
Comment (stebenz) [reply proto/zitadel/settings/v2/login_settings.proto:137]: Oh I think that accidentally went in as I was working on the other PR. Removed.

File: proto/zitadel/saml/v2/saml_service.proto [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+      description: "Token to verify the session is valid";
+    }
+  ];
+}
+
+message CreateCallbackResponse {
+  zitadel.object.v2.Details details = 1;
+  string callback_url = 2 [
+    (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_field) = {
+      description: "Callback URL where the user should be redirected, using a \"302 FOUND\" status. Contains details for the application to obtain the tokens on success, or error details on failure. Note that this field must be treated as credentials, as the contained code can be used to obtain tokens on behalve of the user.";
```
[the remark this answers is not in the document]
Comment (stebenz) [reply proto/zitadel/saml/v2/saml_service.proto:213]: Tried to hold it as much as possible like OIDC, but yes your right, will rename and relable it SAML specific.

File: internal/repository/samlsession/saml_session.go [hunk 3]
```diff
@@ -0,0 +1,141 @@
...
+)
+
+type AddedEvent struct {
+	eventstore.BaseEvent `json:"-"`
+
+	UserID            string                      `json:"userID"`
+	UserResourceOwner string                      `json:"userResourceOwner"`
+	SessionID         string                      `json:"sessionID"`
+	EntityID          string                      `json:"entityID"`
+	Audience          []string                    `json:"audience"`
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/repository/samlsession/saml_session.go:27]: There is an audience in the response, currently only limited to the SP itself, but yes we have one.

File: internal/repository/samlsession/saml_session.go [hunk 4]
```diff
@@ -0,0 +1,141 @@
...
+	}
+}
+
+type SAMLResponseAddedEvent struct {
+	eventstore.BaseEvent `json:"-"`
+
+	ID       string             `json:"id,omitempty"`
+	Lifetime time.Duration      `json:"lifetime,omitempty"`
+	Reason   domain.TokenReason `json:"reason,omitempty"`
+	Actor    *domain.TokenActor `json:"actor,omitempty"`
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/repository/samlsession/saml_session.go:82]: Not used, missed to remove.

File: internal/repository/samlrequest/saml_request.go [hunk 5]
```diff
@@ -0,0 +1,175 @@
...
+	FailedType             = samlRequestEventPrefix + "failed"
+	SessionLinkedType      = samlRequestEventPrefix + "session.linked"
+	SucceededType          = samlRequestEventPrefix + "succeeded"
+)
+
+type AddedEvent struct {
+	*eventstore.BaseEvent `json:"-"`
+
+	LoginClient   string `json:"login_client,omitempty"`
+	ApplicationID string `json:"application_id,omitempty"`
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/repository/samlrequest/saml_request.go:23]: ApplicationID is the application itself, as the entityID could theoratically change, and the entityID is more or less only used to get the applicationID.
We could have implemented it to set the applicationID == entityID, so we would have no unique contraints problems, but that was a decision we took a long time ago.

File: internal/repository/samlrequest/saml_request.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+	ACSURL        string `json:"acs_url,omitempty"`
+	RelayState    string `json:"relay_state,omitempty"`
+	RequestID     string `json:"request_id,omitempty"`
+	Binding       string `json:"binding,omitempty"`
+	Issuer        string `json:"issuer,omitempty"`
+	IssuerName    string `json:"issuer_name,omitempty"`
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/repository/samlrequest/saml_request.go:29]: Only included it as it is included in OIDC

File: internal/repository/samlrequest/saml_request.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+
+type AddedEvent struct {
+	*eventstore.BaseEvent `json:"-"`
+
+	LoginClient   string `json:"login_client,omitempty"`
+	ApplicationID string `json:"application_id,omitempty"`
+	ACSURL        string `json:"acs_url,omitempty"`
+	RelayState    string `json:"relay_state,omitempty"`
+	RequestID     string `json:"request_id,omitempty"`
+	Binding       string `json:"binding,omitempty"`
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/repository/samlrequest/saml_request.go:27]: The binding in itself is a string in the request, and handled there, that's why I did it as a string here. We could parse it if necessary.

File: proto/zitadel/saml/v2/saml_service.proto [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+    option (validate.required) = true;
+    RedirectResponse redirect = 3;
+    PostResponse post = 4;
+  }
+}
+
+message RedirectResponse{}
+message PostResponse{
+  string relay_state = 1;
+  string body = 2;
```
[the remark this answers is not in the document]
Comment (stebenz) [reply proto/zitadel/saml/v2/saml_service.proto:227]: True, yes should always be the response in failure and success.

File: proto/zitadel/saml/v2/saml_service.proto [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+  string callback_url = 2 [
+    (grpc.gateway.protoc_gen_openapiv2.options.openapiv2_field) = {
+      description: "Callback URL where the user should be redirected, using a \"302 FOUND\" status. Contains details for the application to obtain the tokens on success, or error details on failure. Note that this field must be treated as credentials, as the contained code can be used to obtain tokens on behalve of the user.";
+      example: "\"https://client.example.org/cb\""
+    }
+  ];
+  oneof binding {
+    option (validate.required) = true;
+    RedirectResponse redirect = 3;
+    PostResponse post = 4;
```
[the remark this answers is not in the document]
Comment (stebenz) [reply proto/zitadel/saml/v2/saml_service.proto:220]: Structured it this way as you always should have an ACS URL, and you either have a URL with all the information or you have the URL and a post body.
IMO both options would be valid, can't really decide on what feels better.

File: internal/api/saml/auth_request_converter_v2.go [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+}
+func (a *AuthRequestV2) GetRelayState() string {
+	return a.RelayState
+}
+func (a *AuthRequestV2) GetAccessConsumerServiceURL() string {
+	return a.ACSURL
+}
+
+func (a *AuthRequestV2) GetNameID() string {
+	return a.UserID
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/api/saml/auth_request_converter_v2.go:30]: Good point, I think that's something unnecessary, but included in the interface from zitadel/saml, would need to remove it there as well.
https://github.com/zitadel/saml/blob/95c785b0bc2fec76c2901bbded1e685d5a6804fb/pkg/provider/models/models.go#L7

File: internal/api/saml/auth_request_converter_v2.go [hunk 11]
```diff
@@ -0,0 +1,60 @@
...
+	return a.RequestID
+}
+func (a *AuthRequestV2) GetBindingType() string {
+	return a.Binding
+}
+func (a *AuthRequestV2) GetIssuer() string {
+	return a.Issuer
+}
+func (a *AuthRequestV2) GetIssuerName() string {
+	return a.IssuerName
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/api/saml/auth_request_converter_v2.go:43]: Same here 
https://github.com/zitadel/saml/blob/95c785b0bc2fec76c2901bbded1e685d5a6804fb/pkg/provider/models/models.go#L13

File: internal/api/saml/auth_request_converter_v2.go [hunk 12]
```diff
@@ -0,0 +1,60 @@
...
+}
+func (a *AuthRequestV2) GetDestination() string {
+	return a.Destination
+}
+func (a *AuthRequestV2) GetCode() string {
+	return ""
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/api/saml/auth_request_converter_v2.go:49]: Same here https://github.com/zitadel/saml/blob/95c785b0bc2fec76c2901bbded1e685d5a6804fb/pkg/provider/models/models.go#L11

File: internal/api/saml/auth_request_converter_v2.go [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+}
+func (a *AuthRequestV2) GetUserID() string {
+	return a.UserID
+}
+func (a *AuthRequestV2) GetUserName() string {
+	return ""
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/api/saml/auth_request_converter_v2.go:55]: Same here
https://github.com/zitadel/saml/blob/95c785b0bc2fec76c2901bbded1e685d5a6804fb/pkg/provider/models/models.go#L16

File: internal/command/saml_session.go [hunk 14]
```diff
@@ -0,0 +1,186 @@
...
+	"github.com/zitadel/zitadel/internal/telemetry/tracing"
+	"github.com/zitadel/zitadel/internal/zerrors"
+)
+
+type SAMLSession struct {
+	SessionID         string
+	SAMLResponseID    string
+	EntityID          string
+	UserID            string
+	Audience          []string
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/command/saml_session.go:26]: Yes, entityID is used as the audience currently.

File: internal/domain/saml_error_reason.go [hunk 15]
```diff
@@ -0,0 +1,18 @@
...
+	SAMLErrorReasonAuthNFailed
+	SAMLErrorReasonInvalidAttrNameOrValue
+	SAMLErrorReasonInvalidNameIDPolicy
+	SAMLErrorReasonRequestDenied
+	SAMLErrorReasonRequestUnsupported
+	SAMLErrorReasonUnsupportedBinding
+)
+
+func SAMLErrorReasonFromError(err error) SAMLErrorReason {
+	return SAMLErrorReasonUnspecified
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/domain/saml_error_reason.go:17]: Totally overlooked that, was a inbetween solution as I had to move over the relevant errors from zitadel/saml.

File: proto/zitadel/saml/v2/saml_service.proto [hunk 16]
```diff
@@ -0,0 +1,228 @@
...
+      responses: {
+        key: "200"
+        value: {
+          description: "OK";
+        }
+      };
+    };
+  }
+
+  rpc CreateCallback (CreateCallbackRequest) returns (CreateCallbackResponse) {
```
[the remark this answers is not in the document]
Comment (stebenz) [reply proto/zitadel/saml/v2/saml_service.proto:127]: Originally it was to keep it consistent, but if you but it that way in my opinion we should name it something entirely  different.
Maybe just CreateResponse? In artifact binding and other binding you always have a Response message, which contains either the artifact response or the assertion.

File: proto/zitadel/saml/v2/saml_service.proto [hunk 17]
```diff
@@ -0,0 +1,228 @@
...
+        json_schema: {
+          ref: "#/definitions/rpcStatus";
+        }
+      }
+    }
+  }
+};
+
+service SAMLService {
+  rpc GetAuthRequest (GetSAMLRequestRequest) returns (GetSAMLRequestResponse) {
```
[the remark this answers is not in the document]
Comment (stebenz) [reply proto/zitadel/saml/v2/saml_service.proto:104]: Done

File: internal/api/saml/auth_request.go [hunk 18]
```diff
@@ -0,0 +1,101 @@
...
+
+	switch authReq.GetBindingType() {
+	case provider.PostBinding:
+		return authReq.GetAccessConsumerServiceURL(), base64.StdEncoding.EncodeToString(respData), nil
+	case provider.RedirectBinding:
+		respData, err := xml.DeflateAndBase64(respData)
+		if err != nil {
+			return "", "", err
+		}
+		return fmt.Sprintf("%s?%s", authReq.GetAccessConsumerServiceURL(), provider.BuildRedirectQuery(string(respData), resp.RelayState, resp.SigAlg, resp.Signature)), "", nil
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/api/saml/auth_request.go:40]: Not quite sure here, if I recall correctly there could be SAML entities using query params for resolving the enitityID, should I at least clear the specific SAML parameters?

File: internal/command/saml_session_model.go [hunk 19]
```diff
@@ -0,0 +1,101 @@
...
+	// the write model might be initialized without resource owner,
+	// so update the aggregate
+	if wm.ResourceOwner == "" {
+		wm.aggregate = &samlsession.NewAggregate(wm.AggregateID, e.Aggregate().ResourceOwner).Aggregate
+	}
+}
+
+func (wm *SAMLSessionWriteModel) reduceSAMLResponseAdded(e *samlsession.SAMLResponseAddedEvent) {
+	wm.SAMLResponseID = e.ID
+	wm.SAMLResponseCreation = e.CreationDate().Add(e.Lifetime)
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/command/saml_session_model.go:95]: As mentioned in our conversion, left out for now as it is mainly used in OIDC impersonation.

File: internal/api/saml/auth_request.go [hunk 20]
```diff
@@ -0,0 +1,101 @@
...
+	}
+	return "", "", nil
```
[the remark this answers is not in the document]
Comment (stebenz) [reply internal/api/saml/auth_request.go]: moved it to a separate function

**Review by livio-a — CHANGES_REQUESTED** [review]

**Review by livio-a — APPROVED** [review]

**Comment by stebenz** [comment]
Issue for documentation added https://github.com/zitadel/zitadel/issues/9088

**Pull request merged**